plugins/env/app/decorators/deploy_decorator.rb

```diff
@@ -13,7 +13,18 @@ def serialized_environment_variables
     variables = EnvironmentVariable.nested_variables(project)
 
     if deploy_groups = stage.deploy_groups.presence
-      variables = variables.select { |ev| deploy_groups.any? { |dg| ev.matches_scope?(dg) } }
+      # filter variables to only those matching stage's deploy groups and environments, includes 'All'
+      variables.select! { |ev| deploy_groups.any? { |dg| ev.matches_scope?(dg) } }
+
+      # sort by priority (Deploy Groups > Environment > All)
+      variables.sort_by!(&:priority)
+
+      # take all deploy group variables of the same name since all would be used in the deploy
+      # if no deploy groups variables are present, take the variable of the highest precedence
+      variables = variables.group_by(&:name).flat_map do |_name, env_vars|
+        deploy_group_env_vars, other_env_vars = env_vars.partition { |env_var| env_var.scope_type == 'DeployGroup' }
+        deploy_group_env_vars.present? ? deploy_group_env_vars : other_env_vars.first
+      end
```

Comment:
I think originally I didn't use it because of [performance concerns](https://github.com/zendesk/samson/pull/2664#discussion_r178201563). Now that's not really an issue because we need to do that work anyway. I could use something like `SamsonEnv.env_groups` and then make my own serialization method?

Reply:
My thinking is that as much as possible, we want to use the same logic for the preview as we do to determine the actual values. That way, we maximize the likelihood they stay in sync.